### DISCUSSION TITLE
Dependency Injection vs Middleware

### DISCUSSION THREAD

**Discussion by assman** [fastapi/fastapi#8867] [Questions] [answered]
I am wondering what is the recommended way to go about adding authentication to path operations? 

FastAPI allows us to use dependency injection or middleware for this task but I was wondering if anyone had any takes on which is a better way overall.

**Comment by dstlny** [comment]
> I am wondering what is the recommended way to go about adding authentication to path operations?
> 
> FastAPI allows us to use dependency injection or middleware for this task but I was wondering if anyone had any takes on which is a better way overall.

You could always use the @requires decorator from starlette, along with Authentication Middleware. Say, a authenticated user has a role of 'system:authenticated', that way you'd use the @requires decorator on whichever paths you wanted secured to only authenticated users with that role)

@requires decorator: https://github.com/encode/starlette/blob/d26f44b4aeaec63ae8ce0a4f34dea7439da48012/starlette/authentication.py#L19

Starlette documentation on Middleware + @requires decorator: https://www.starlette.io/authentication/#permissions

**Comment by DavianYang** [comment]
You can also try or reference from [fastapi-permission](https://github.com/holgi/fastapi-permissions)

**Comment by Kludex** [comment, marked as answer]
@DavianYang What? hahaha No no no, FastAPI is fine. 

> I am wondering what is the recommended way to go about adding authentication to path operations?
> 
> FastAPI allows us to use dependency injection or middleware for this task but I was wondering if anyone had any takes on which is a better way overall.

@assman Usually, you have a dependency taking care of it. You can take [this project](https://github.com/tiangolo/full-stack-fastapi-postgresql/tree/master/%7B%7Bcookiecutter.project_slug%7D%7D/backend/app/app) to have an idea.

**Reply by Kludex** [reply]
> What? hahaha No no no, FastAPI is fine.

Kludex from the past was too excited on the answer.

Use whatever you feel like. 👍

**Comment by DavianYang** [comment]
@Kludex, Sry I gave him wrong link since [Pyramid](https://trypyramid.com/) is reference for that repo

**Comment by assman** [comment]
> @DavianYang What? hahaha No no no, FastAPI is fine. 
> 
> 
> 
> > I am wondering what is the recommended way to go about adding authentication to path operations?
> 
> > 
> 
> > FastAPI allows us to use dependency injection or middleware for this task but I was wondering if anyone had any takes on which is a better way overall.
> 
> 
> 
> @assman Usually, you have a dependency taking care of it. You can take [this project](https://github.com/tiangolo/full-stack-fastapi-postgresql/tree/master/%7B%7Bcookiecutter.project_slug%7D%7D/backend/app/app) to have an idea.

Ok noted thank you

**Comment by gelycot** [comment]
It is advisable to use the approach that best suits your needs and the architecture of your project. Both methods offer different levels of flexibility and control for implementing authentication. Middleware functions as an intermediary layer that processes requests before they reach your application while another approach facilitates the injection of dependencies, such as authentication objects, into functions or methods.

**Reply by assman** [reply]
Makes sense, thanks for the input.

**Comment by hekimazuri** [comment]
Apologies for necro-bumping, but my team is having exactly this debate right now.

Given that we have a huge number of routes and all but one or two require auth, we though it would be simpler to use a middleware.

@Kludex You said usually dependency injection is used, but would it be considered a bad practice to use a middleware/are there downsides to using a middleware over dependency injection?

**Reply by assman** [reply]
What did you end up using in the end?
